In [329]:
import pandas as pd

In [330]:
X_train = pd.read_csv('X_train.csv', encoding='CP949')
X_train


cust_id       총구매액     최대구매액       환불금액   주구매상품 주구매지점  내점일수   내점당구매건수  \
0           0   68282840  11264000  6860000.0      기타   강남점    19  3.894737   
1           1    2136000   2136000   300000.0     스포츠   잠실점     2  1.500000   
2           2    3197000   1639000        NaN  남성 캐주얼   관악점     2  2.000000   
3           3   16077620   4935000        NaN      기타   광주점    18  2.444444   
4           4   29050000  24000000        NaN      보석  본  점     2  1.500000   
...       ...        ...       ...        ...     ...   ...   ...       ...   
3495     3495    3175200   3042900        NaN      골프  본  점     1  2.000000   
3496     3496   29628600   7200000  6049600.0    시티웨어  부산본점     8  1.625000   
3497     3497      75000     75000        NaN    주방용품   창원점     1  1.000000   
3498     3498    1875000   1000000        NaN     화장품  본  점     2  1.000000   
3499     3499  263101550  34632000  5973000.0      기타  본  점    38  2.421053   

        주말방문비율  구매주기  
0     0.527027    17  
1     0.000000     1  
2     0.000000     1  
3     0.318182    16  
4     0.000000    85  
...        ...   ...  
3495  1.000000     0  
3496  0.461538    40  
3497  0.000000     0  
3498  0.000000    39  
3499  0.467391     8  

[3500 rows x 10 columns]

In [331]:
X_test = pd.read_csv('X_test.csv', encoding='CP949')
y_train = pd.read_csv('y_train.csv', encoding='CP949')
X_test


cust_id       총구매액     최대구매액        환불금액   주구매상품 주구매지점  내점일수    내점당구매건수  \
0        3500   70900400  22000000   4050000.0      골프  부산본점    13   1.461538   
1        3501  310533100  38558000  48034700.0     농산물   잠실점    90   2.433333   
2        3502  305264140  14825000  30521000.0    가공식품  본  점   101  14.623762   
3        3503    7594080   5225000         NaN    주방용품  부산본점     5   2.000000   
4        3504    1795790   1411200         NaN     수산품  청량리점     3   2.666667   
...       ...        ...       ...         ...     ...   ...   ...        ...   
2477     5977   82581500  23976000         NaN      골프  부산본점     8   1.750000   
2478     5978     480000    480000         NaN    섬유잡화   광주점     1   1.000000   
2479     5979  260003790  25750000         NaN  남성 캐주얼  본  점    19   3.736842   
2480     5980   88991520  18120000         NaN      육류  본  점     5   3.600000   
2481     5981     623700    209000         NaN    가공식품  영등포점     2   5.000000   

        주말방문비율  구매주기  
0     0.789474    26  
1     0.369863     3  
2     0.083277     3  
3     0.000000    47  
4     0.125000     8  
...        ...   ...  
2477  0.642857    40  
2478  0.000000     0  
2479  0.915493    18  
2480  0.444444    60  
2481  0.000000    31  

[2482 rows x 10 columns]

In [332]:
y_train

cust_id  gender
0           0       0
1           1       0
2           2       1
3           3       1
4           4       0
...       ...     ...
3495     3495       1
3496     3496       1
3497     3497       0
3498     3498       0
3499     3499       0

[3500 rows x 2 columns]

In [333]:
# X_train 에서 cust_id 컬럼 삭제하기
# X_train = X_train.drop(['cust_id'], axis=1, inplace=False)
# X_train

In [334]:
X_train = X_train.iloc[:,1:]
X_train

총구매액     최대구매액       환불금액   주구매상품 주구매지점  내점일수   내점당구매건수    주말방문비율  \
0      68282840  11264000  6860000.0      기타   강남점    19  3.894737  0.527027   
1       2136000   2136000   300000.0     스포츠   잠실점     2  1.500000  0.000000   
2       3197000   1639000        NaN  남성 캐주얼   관악점     2  2.000000  0.000000   
3      16077620   4935000        NaN      기타   광주점    18  2.444444  0.318182   
4      29050000  24000000        NaN      보석  본  점     2  1.500000  0.000000   
...         ...       ...        ...     ...   ...   ...       ...       ...   
3495    3175200   3042900        NaN      골프  본  점     1  2.000000  1.000000   
3496   29628600   7200000  6049600.0    시티웨어  부산본점     8  1.625000  0.461538   
3497      75000     75000        NaN    주방용품   창원점     1  1.000000  0.000000   
3498    1875000   1000000        NaN     화장품  본  점     2  1.000000  0.000000   
3499  263101550  34632000  5973000.0      기타  본  점    38  2.421053  0.467391   

      구매주기  
0       17  
1        1  
2        1  
3       16  
4       85  
...    ...  
3495     0  
3496    40  
3497     0  
3498    39  
3499     8  

[3500 rows x 9 columns]

In [335]:
X_test

cust_id       총구매액     최대구매액        환불금액   주구매상품 주구매지점  내점일수    내점당구매건수  \
0        3500   70900400  22000000   4050000.0      골프  부산본점    13   1.461538   
1        3501  310533100  38558000  48034700.0     농산물   잠실점    90   2.433333   
2        3502  305264140  14825000  30521000.0    가공식품  본  점   101  14.623762   
3        3503    7594080   5225000         NaN    주방용품  부산본점     5   2.000000   
4        3504    1795790   1411200         NaN     수산품  청량리점     3   2.666667   
...       ...        ...       ...         ...     ...   ...   ...        ...   
2477     5977   82581500  23976000         NaN      골프  부산본점     8   1.750000   
2478     5978     480000    480000         NaN    섬유잡화   광주점     1   1.000000   
2479     5979  260003790  25750000         NaN  남성 캐주얼  본  점    19   3.736842   
2480     5980   88991520  18120000         NaN      육류  본  점     5   3.600000   
2481     5981     623700    209000         NaN    가공식품  영등포점     2   5.000000   

        주말방문비율  구매주기  
0     0.789474    26  
1     0.369863     3  
2     0.083277     3  
3     0.000000    47  
4     0.125000     8  
...        ...   ...  
2477  0.642857    40  
2478  0.000000     0  
2479  0.915493    18  
2480  0.444444    60  
2481  0.000000    31  

[2482 rows x 10 columns]

In [336]:
X_test_id = X_test['cust_id']

In [337]:
# X_test 에서 cust_id 삭제
X_test = X_test.drop(['cust_id'], axis=1, inplace=False)
X_test

총구매액     최대구매액        환불금액   주구매상품 주구매지점  내점일수    내점당구매건수  \
0      70900400  22000000   4050000.0      골프  부산본점    13   1.461538   
1     310533100  38558000  48034700.0     농산물   잠실점    90   2.433333   
2     305264140  14825000  30521000.0    가공식품  본  점   101  14.623762   
3       7594080   5225000         NaN    주방용품  부산본점     5   2.000000   
4       1795790   1411200         NaN     수산품  청량리점     3   2.666667   
...         ...       ...         ...     ...   ...   ...        ...   
2477   82581500  23976000         NaN      골프  부산본점     8   1.750000   
2478     480000    480000         NaN    섬유잡화   광주점     1   1.000000   
2479  260003790  25750000         NaN  남성 캐주얼  본  점    19   3.736842   
2480   88991520  18120000         NaN      육류  본  점     5   3.600000   
2481     623700    209000         NaN    가공식품  영등포점     2   5.000000   

        주말방문비율  구매주기  
0     0.789474    26  
1     0.369863     3  
2     0.083277     3  
3     0.000000    47  
4     0.125000     8  
...        ...   ...  
2477  0.642857    40  
2478  0.000000     0  
2479  0.915493    18  
2480  0.444444    60  
2481  0.000000    31  

[2482 rows x 9 columns]

In [338]:
# y_train 에서 cust_id 삭제
y_train = y_train.iloc[:,-1:]
y_train

gender
0          0
1          0
2          1
3          1
4          0
...      ...
3495       1
3496       1
3497       0
3498       0
3499       0

[3500 rows x 1 columns]

In [339]:
X_train.loc[:,['주구매상품','주구매지점']]

주구매상품 주구매지점
0         기타   강남점
1        스포츠   잠실점
2     남성 캐주얼   관악점
3         기타   광주점
4         보석  본  점
...      ...   ...
3495      골프  본  점
3496    시티웨어  부산본점
3497    주방용품   창원점
3498     화장품  본  점
3499      기타  본  점

[3500 rows x 2 columns]

In [340]:
from sklearn.preprocessing import LabelEncoder
# X_train
# \는 2개가 연결되었다는 의미
X_train.loc[:,['주구매상품','주구매지점']] = \
X_train.loc[:,['주구매상품','주구매지점']].apply(LabelEncoder().fit_transform)

In [341]:
X_train.loc[:,['주구매상품','주구매지점']]

주구매상품  주구매지점
0         5      0
1        21     19
2         6      1
3         5      2
4        15      8
...     ...    ...
3495      3      8
3496     22      9
3497     32     21
3498     41      8
3499      5      8

[3500 rows x 2 columns]

In [342]:
# X_test LabelEncoder 적용 
X_test.loc[:,['주구매상품','주구매지점']] = \
X_test.loc[:,['주구매상품','주구매지점']].apply(LabelEncoder().fit_transform)

In [343]:
X_test.loc[:,['주구매상품','주구매지점']]

주구매상품  주구매지점
0         3      9
1         9     19
2         0      8
3        31      9
4        19     22
...     ...    ...
2477      3      9
2478     17      2
2479      6      8
2480     26      8
2481      0     15

[2482 rows x 2 columns]

In [344]:
#null 확인
X_train.isnull().sum()

총구매액          0
최대구매액         0
환불금액       2295
주구매상품         0
주구매지점         0
내점일수          0
내점당구매건수       0
주말방문비율        0
구매주기          0
dtype: int64

In [345]:
# X_train['환불금액'] = X_train['환불금액'].fillna(0)
# X_train.isnull().sum()

In [346]:
#loc() 사용
X_train.loc[:,['환불금액']] = X_train.loc[:,['환불금액']].fillna(0)
X_train.isnull().sum()

총구매액       0
최대구매액      0
환불금액       0
주구매상품      0
주구매지점      0
내점일수       0
내점당구매건수    0
주말방문비율     0
구매주기       0
dtype: int64

In [347]:
X_test.isnull().sum()

총구매액          0
최대구매액         0
환불금액       1611
주구매상품         0
주구매지점         0
내점일수          0
내점당구매건수       0
주말방문비율        0
구매주기          0
dtype: int64

In [348]:
X_test['환불금액'] = X_test['환불금액'].fillna(0)
X_test.isnull().sum()

총구매액       0
최대구매액      0
환불금액       0
주구매상품      0
주구매지점      0
내점일수       0
내점당구매건수    0
주말방문비율     0
구매주기       0
dtype: int64

In [349]:
# 로지스틱 회귀분석 <--- qnsfb
from sklearn.linear_model import LogisticRegression
import warnings
warnings.filterwarnings('ignore')

md = LogisticRegression(random_state=0)
md.fit(X_train, y_train)
print('xscore : ', md.score(X_train, y_train))

xscore :  0.624


작업형 2. 아래는 백화점 고객의 1년 간 구매데이터이다. 고객 3500명에 대한 학습용(y_train.csv, X_train.csv) 데이터를 이용하여 성별예측 모형을 만든 후, 이를 평가용데이터(X_test.csv)에 적용하여 얻은 2482명 고객의 성별 예측값(남자일 확률)을 다음과 같은 형식(custid, gender) 의 CSV 파일로 생성하시오.

In [350]:
# predict_proba : 해당 레이블로 분류될 확률 값
pred = md.predict_proba(X_test)
pred = pd.DataFrame(pred)
pred = pred.iloc[:,1]
pred

0       0.425232
1       0.198986
2       0.238942
3       0.490611
4       0.497643
          ...   
2477    0.423941
2478    0.499290
2479    0.307654
2480    0.423974
2481    0.499399
Name: 1, Length: 2482, dtype: float64

In [351]:
# X_test['gender'] = pred
# X_test

In [352]:
# df = pd.read_csv('X_test.csv', encoding='CP949')
# df['gender'] = pred
# answer = df.loc[:,['cust_id','gender']]
# answer
# answer.to_csv('answer.csv')

In [353]:
answer = pd.concat([X_test_id,pred],axis=1)
answer.to_csv('gender.csv')

In [354]:
pd.read_csv('gender.csv')

Unnamed: 0  cust_id         1
0              0     3500  0.425232
1              1     3501  0.198986
2              2     3502  0.238942
3              3     3503  0.490611
4              4     3504  0.497643
...          ...      ...       ...
2477        2477     5977  0.423941
2478        2478     5978  0.499290
2479        2479     5979  0.307654
2480        2480     5980  0.423974
2481        2481     5981  0.499399

[2482 rows x 3 columns]

In [355]:
# K-최근접 이웃 알고리즘
from sklearn.neighbors import KNeighborsClassifier

md = KNeighborsClassifier(n_neighbors=10, metric= 'euclidean')
md

KNeighborsClassifier(metric='euclidean', n_neighbors=10)

In [356]:
md.fit(X_train, y_train)
print('score : ', md.score(X_train, y_train))

score :  0.6871428571428572


In [357]:
pred = md.predict_proba(X_test)
pred = pd.DataFrame(pred)
pred = pred.iloc[:,1]
pred

0       0.2
1       0.2
2       0.1
3       0.4
4       0.3
       ... 
2477    0.6
2478    0.6
2479    0.2
2480    0.4
2481    0.6
Name: 1, Length: 2482, dtype: float64

In [358]:
ans = pd.concat([X_test_id,pred], axis=1)
ans

cust_id    1
0        3500  0.2
1        3501  0.2
2        3502  0.1
3        3503  0.4
4        3504  0.3
...       ...  ...
2477     5977  0.6
2478     5978  0.6
2479     5979  0.2
2480     5980  0.4
2481     5981  0.6

[2482 rows x 2 columns]